src/memray/_memray/record_writer.cpp

```diff
@@ -93,23 +392,266 @@ RecordWriter::writeTrailer()
+            frame_id,
+            Frame{raw.function_name, raw.filename, raw.lineno, raw.is_entry_frame});
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeRecord(const UnresolvedNativeFrame& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_native_frames.emplace_back(record);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeMappings(const std::vector<ImageSegments>& mappings)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_mappings_by_generation.push_back(mappings);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const FramePop& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    auto count = record.count;
+    auto& stack = d_python_stack_ids_by_thread[tid];
+    assert(stack.size() >= record.count);
+    while (count) {
+        count -= 1;
+        stack.pop_back();
+    }
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const FramePush& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    auto [it, inserted] = d_python_stack_ids_by_thread.emplace(tid, python_stack_ids_t{});
+    auto& stack = it->second;
+    if (inserted) {
+        stack.reserve(1024);
+    }
+    FrameTree::index_t current_stack_id = stack.empty() ? 0 : stack.back();
+    FrameTree::index_t new_stack_id =
+            d_python_frame_tree.getTraceIndex(current_stack_id, record.frame_id);
+    stack.push_back(new_stack_id);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const AllocationRecord& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    Allocation allocation;
+    allocation.tid = tid;
+    allocation.address = record.address;
+    allocation.size = record.size;
+    allocation.allocator = record.allocator;
+    allocation.native_frame_id = 0;
+    if (!hooks::isDeallocator(record.allocator)) {
+        auto& stack = d_python_stack_ids_by_thread[tid];
+        allocation.frame_index = stack.empty() ? 0 : stack.back();
+    } else {
+        allocation.frame_index = 0;
+    }
+    allocation.native_segment_generation = 0;
+    allocation.n_allocations = 1;
+    d_high_water_mark_aggregator.addAllocation(allocation);
+    return true;
+}
+
+bool
+AggregatingRecordWriter::writeThreadSpecificRecord(thread_id_t tid, const NativeAllocationRecord& record)
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    Allocation allocation;
+    allocation.tid = tid;
+    allocation.address = record.address;
+    allocation.size = record.size;
+    allocation.allocator = record.allocator;
+    allocation.native_frame_id = record.native_frame_id;
+    auto& stack = d_python_stack_ids_by_thread[tid];
+    allocation.frame_index = stack.empty() ? 0 : stack.back();
+    allocation.native_segment_generation = d_mappings_by_generation.size();
+    allocation.n_allocations = 1;
+    d_high_water_mark_aggregator.addAllocation(allocation);
```

Comment:
Same as before, any reason why we are not just doing

```suggestion
    std::lock_guard<std::mutex> lock(d_mutex);
    auto& stack = d_python_stack_ids_by_thread[tid];
    Allocation allocation{
     tid,
     record.address,
     record.size,
    record.allocator,
    record.native_frame_id,
    stack.empty() ? 0 : stack.back(),
    d_mappings_by_generation.size(),
    };
    d_high_water_mark_aggregator.addAllocation(allocation);
```

Reply:
I prefer not to use positional struct initialization for structures with more than 3 or 4 fields, as we've seen that it's a frequent source of bugs where a value gets assigned to the wrong field accidentally. Your proposal is the same number of lines of code, and with your version someone reading the code can't be sure whether the correct parameter was assigned to each field without opening the header and cross-referencing the initialization against the order of fields in the `struct` declaration. Same reason as I try not to pass more than 3 or 4 positional parameters in Python code.